src/server/middleware/immutable/deletemf.go

```diff
@@ -0,0 +1,102 @@
+package immutable
+
+import (
+	"errors"
+	"fmt"
+	common_util "github.com/goharbor/harbor/src/common/utils"
+	internal_errors "github.com/goharbor/harbor/src/internal/error"
+	"github.com/goharbor/harbor/src/pkg/art"
+	"github.com/goharbor/harbor/src/pkg/artifact"
+	"github.com/goharbor/harbor/src/pkg/immutabletag/match/rule"
+	"github.com/goharbor/harbor/src/pkg/q"
+	"github.com/goharbor/harbor/src/pkg/repository"
+	"github.com/goharbor/harbor/src/pkg/tag"
+	"github.com/goharbor/harbor/src/server/middleware"
+	"net/http"
+)
+
+// MiddlewareDelete ...
+func MiddlewareDelete() func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			if err := handleDelete(req); err != nil {
+				var e *ErrImmutable
+				if errors.As(err, &e) {
+					pkgE := internal_errors.New(e).WithCode(internal_errors.PreconditionCode)
+					msg := internal_errors.NewErrs(pkgE).Error()
+					http.Error(rw, msg, http.StatusPreconditionFailed)
+					return
+				}
+				pkgE := internal_errors.New(fmt.Errorf("error occurred when to handle request in immutable handler: %v", err)).WithCode(internal_errors.GeneralCode)
+				msg := internal_errors.NewErrs(pkgE).Error()
+				http.Error(rw, msg, http.StatusInternalServerError)
+			}
+			next.ServeHTTP(rw, req)
+		})
+	}
+}
+
+// handleDelete ...
+func handleDelete(req *http.Request) error {
+	mf, ok := middleware.ManifestInfoFromContext(req.Context())
+	if !ok {
+		return errors.New("cannot get the manifest information from request context")
+	}
+
+	_, repoName := common_util.ParseRepository(mf.Repository)
+	total, repos, err := repository.Mgr.List(req.Context(), &q.Query{
+		Keywords: map[string]interface{}{
+			"Name": mf.Repository,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if total == 0 {
+		return nil
+	}
+
+	total, afs, err := artifact.Mgr.List(req.Context(), &q.Query{
```

Comment:
Should call the list of artifact controller, it returns all the tags and immutable status. As the immutable status isn't populated into the returning result of the function yet, open an issue https://github.com/goharbor/harbor/issues/10545 to track the refactor work.